etcd-backup/main.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/spf13/cobra"
+	"go.etcd.io/etcd/client/pkg/v3/transport"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/snapshot"
+
+	"go.uber.org/zap"
+)
+
+const (
+	DefaultDialTimeout       = 15 * time.Second
+	DefaultEtcdClientTimeout = 5 * time.Minute
+)
+
+type options struct {
+	clusterID      string
+	etcdEndpoint   string
+	s3BucketName   string
+	s3BucketRegion string
+}
+
+func main() {
+	var opts options
+
+	cmd := &cobra.Command{
+		Use:          "etcd-backup",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return run(opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.clusterID, "cluster-id", "", "ID of the hosted cluster, used as a key for the snapshot file in S3")
+	cmd.Flags().StringVar(&opts.etcdEndpoint, "etcd-endpoint", "", "endpoint of the etcd cluster to backup.")
+	cmd.Flags().StringVar(&opts.s3BucketName, "s3-bucket-name", "", "name of the S3 bucket to store etcd backups.")
+	cmd.Flags().StringVar(&opts.s3BucketRegion, "s3-bucket-region", "", "AWS region of the S3 bucket to store etcd backups.")
```

Comment:
It would also be good to specify the directory where to store the local backup, and when we run this in a pod, mount an emptydir at that location. I don't think it's good to assume that /tmp will always be writable